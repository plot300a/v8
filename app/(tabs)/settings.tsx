
import SettingsScreen from "@/screens/settings";

export default function Setting() {
   
    return <SettingsScreen />
}

