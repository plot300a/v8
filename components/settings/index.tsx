import {
    ImageStyle,
    ImageURISource,
    StyleSheet,
    TextInput,
    TextStyle,
    TouchableOpacity,
    View,
    ViewStyle,
} from "react-native";
import React, { useMemo } from "react";
import { Ionicons } from "@expo/vector-icons";
import { Text, useThemeColor } from "@/theme";
import { fontSize } from "@/constants/FontSizes";
import { CustomImageComponent } from "@/commons/Image";
import { Colors } from "@/constants/Colors";

export const TabHead = ({ title, active }: { title: string; active?: boolean }) => {
    const { dark, gray1, primary } = useThemeColor();
    const style = useMemo(() => {
        return {
            tabHeadText: (active
                ? {
                    ...styles.tabHeadText,
                    color: dark,
                }
                : {
                    ...styles.tabHeadText,
                    color: gray1,
                }),
                indicator: active?{backgroundColor: primary, height: 3, width: '50%'}:{},

        };
    }, [active]);
    return (
        <TouchableOpacity style={styles.tabHeadContainer}>
            <Text testID="tab-title" style={style.tabHeadText}>{title}</Text>
            {active&&<View style={style.indicator as any}></View>}
        </TouchableOpacity>
    );
};

export const InfoInput = ({
    multiline,
    initValue,
    name,
}: {
    multiline?: boolean;
    initValue?: string;
    name: string;
}) => {
    const { gray2, gray0, dark } = useThemeColor();
    const style = useMemo(() => {
        return {
            infoNameText: {
                color: gray2,
                fontWeight: 'bold',
                marginBottom: 5,
            },
            infonInputBottomBorder: {
                height: 1,
                backgroundColor: gray0,
                width: '100%'
            },
            infonInput:[
                styles.infonInput,
                {
                    color: dark
                }
            ]
        };
    }, [dark]);

    initValue = initValue || "";
    return (
        <View testID={`input-field-${name}`} accessible={true} accessibilityLabel={`Edit ${name}`} style={styles.infoInputContainer}>
            <Text style={style.infoNameText as any}>{name}</Text>
            <TextInput
                multiline={multiline}
                value={initValue}
                style={style.infonInput}
                
            />
            <View style={style.infonInputBottomBorder as any}></View>
        </View>
    );
};

type ProfileInfoProps = {
    username: string;
    location: string;
    bio?: string;
    website?: string;
    folllowing: number;
    followers: number;
    imageUrl?: string;
    devImage?: ImageURISource;
};
export const ProfileInfo = ({
    username,
    imageUrl,
    devImage,
    website,
    location,
    folllowing,
    followers,
    bio,
}: ProfileInfoProps) => {
    const { lightPrimary, gray0, icon, primary } = useThemeColor();
    const style:
        | StyleSheet.NamedStyles<any>
        | { [k: string]: (ViewStyle | TextStyle | ImageStyle)[] } = useMemo(() => {
            return {
                searchButtonContainer: {
                    ...styles.searchButtonContainer,
                    backgroundColor: lightPrimary,
                },
                profileImageContainer: {
                    ...styles.profileImageContainer,
                    backgroundColor: gray0,
                },
            };
        }, [gray0]);
    return (
        <View>
            <View style={styles.firstLine}>
                <View style={style.profileImageContainer}>
                    <CustomImageComponent uri={imageUrl || ""} devImage={devImage} imageAlt={`${username}'s profile image`} />
                </View>
                <TouchableOpacity style={style.searchButtonContainer}>
                    <Ionicons name="search" size={20} color={primary} />
                </TouchableOpacity>
            </View>
            <View style={styles.secondLine}>
                <Text style={styles.nameStyle}>{username}</Text>
                <Text style={styles.bioText}>{bio}</Text>
                <View style={styles.pointedInfo}>
                    <View style={styles.pointedContainer}>
                        <Ionicons name="location" size={16} color={icon} />
                        <Text style={styles.pointedText}>{location}</Text>
                    </View>
                    <TouchableOpacity style={styles.pointedContainer}>
                        <Ionicons name="link-outline" size={16} color={icon} />
                        <Text style={styles.pointedLink}>{website}</Text>
                    </TouchableOpacity>
                </View>
                <View style={styles.followInfo}>
                    <View style={styles.followItem}>
                        <Text style={styles.followCount}>{followers}</Text>
                        <Text>Followers</Text>
                    </View>
                    <View style={styles.followItem}>
                        <Text style={styles.followCount}>{folllowing}</Text>
                        <Text>Following</Text>
                    </View>
                </View>
            </View>
        </View>
    );
};

export const commonStyles = StyleSheet.create({
    alignRow: {
        display: "flex",
        alignItems: "center",
        flexDirection: "row",
    },
    alignColumn: {
        display: "flex",
        alignItems: "center",
        flexDirection: "column",
    },
});

export const styles = StyleSheet.create({
    searchButtonContainer: {
        ...commonStyles.alignRow,
        justifyContent: "center",
        padding: 5,
        borderRadius: 30,
    },
    firstLine: {
        ...commonStyles.alignRow,
        justifyContent: "space-between",
        marginBottom: 15,
    },
    profileImageContainer: {
        width: 62,
        height: 62,
        borderRadius: 20,
    },
    secondLine: {
        width: "90%",
        marginBottom: 15,
    },
    nameStyle: {
        fontWeight: "bold",
        fontSize: fontSize.large,
        marginBottom: 5,
    },
    bioText: {
        fontSize: fontSize.normal,
        marginBottom: 5,
        fontWeight: '500'
    },
    pointedInfo: {
        flexDirection: "row",
        flexWrap: "wrap",
        marginBottom: 15,
        marginTop: 5
    },
    pointedContainer: {
        ...commonStyles.alignRow,
        marginRight: 10,
    },
    pointedText: {
        fontSize: fontSize.normal,
        marginLeft: 2,
    },
    pointedLink: {
        fontSize: fontSize.normal,
        marginLeft: 2,
        color: Colors.link,
    },
    followInfo: {
        ...commonStyles.alignRow,
        marginBottom: 15,
    },
    followItem: {
        marginRight: 25,
    },
    followCount: {
        fontWeight: "bold",
        fontSize: 15,
    },
    tabHeadScrollContainer: {
        paddingLeft: 10,
        paddingRight: 10,
        paddingVertical: 5,
        marginBottom: 10
    },
    tabHeadContainer: {
        ...commonStyles.alignColumn,
        marginRight: 15,
        marginLeft: -10,
    },
    tabHeadText: {
        fontWeight: "bold",
        fontSize: 15,
        paddingBottom: 5,
    },
    infoInputContainer: {
        marginBottom: 25,
    },
    infonInputNameText: {},
    infonInput: {
        backgroundColor: "rgba(0,0,0,0)",
        width: "100%",
        marginTop: 10,
        marginBottom: 3,
        fontSize: fontSize.normal,
        paddingBottom: 5,
    },
});
