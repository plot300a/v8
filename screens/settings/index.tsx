import {
    ImageStyle,
    ImageURISource,
    Pressable,
    ScrollView,
    StyleSheet,
    TextInput,
    TextStyle,
    TouchableOpacity,
    View,
    ViewStyle,
} from "react-native";
import React, { useMemo } from "react";
import MainContainer from "@/components/MainContainer";
import CustomHeader from "@/commons/CustomHeader";
import { Text, useThemeColor } from "@/theme";
import { fontSize } from "@/constants/FontSizes";
import { CustomImageComponent } from "@/commons/Image";
import { InfoInput, ProfileInfo, TabHead, styles, commonStyles } from "@/components/settings";

const devUserImage = require('../../assets/images/user.png')

export default function SettingsScreen() {
    const { lightPrimary, gray0, icon, primary } = useThemeColor();
    const style:
        | StyleSheet.NamedStyles<any>
        | { [k: string]: (ViewStyle | TextStyle | ImageStyle)[] } = useMemo(() => {
            return {
                profileImageContainer: {
                    ...styles.profileImageContainer,
                    backgroundColor: gray0,
                },
                saveButtonContainer: {
                    ...commonStyles.alignRow,
                    paddingHorizontal: 10,
                    paddingVertical: 7,
                    borderRadius: 15,
                    borderColor: primary,
                    borderWidth: 1,
                },
                saveText: {
                    color: primary,
                    fontSize: fontSize.normal,
                },
                imageTop:{
                    position:'absolute',top:0,
                    bottom:0,right:0,left:0,
                    backgroundColor:lightPrimary,
                    borderRadius: 20
                },
            };
        }, [gray0]);

    // Some needed data
    const imageUrl = "",
    devImage = devUserImage,
    username="Alexander Moore",
    bio="Welcome to the official account of the world's largest music streaming hub.",
    website="www.versess.com",
    followers=534,
    folllowing=54,
    location="Accra, Ghana";

    return (
        <MainContainer>
            <CustomHeader title="Settings" />
            <ScrollView showsVerticalScrollIndicator={false}>
                <ProfileInfo
                    username={username}
                    bio={bio}
                    imageUrl={imageUrl}
                    website={website}
                    followers={followers}
                    folllowing={folllowing}
                    location={location}
                    devImage={devImage}
                />

                <ScrollView
                    contentContainerStyle={styles.tabHeadScrollContainer}
                    horizontal={true}
                    showsHorizontalScrollIndicator={false}
                >
                    <TabHead title="Posts" active={false} />
                    <TabHead title="Profile Details" active={!false} />
                    <TabHead title="Payment Method" active={false} />
                    <TabHead title="Subscription" active={false} />
                    <TabHead title="Help" active={false} />
                </ScrollView>
                <View>
                    <View style={styles.firstLine}>
                        <View style={style.profileImageContainer}>
                            <CustomImageComponent uri={imageUrl || ""} devImage={devImage} style={{opacity: 0.5}} />
                            <Pressable testID="profile-image-update-button" style={style.imageTop}>

                            </Pressable>
                        </View>
                        <TouchableOpacity testID="save-button" style={style.saveButtonContainer}>
                            <Text style={style.saveText}>Save</Text>
                        </TouchableOpacity>
                    </View>

                    <View>
                        <InfoInput name="Name" initValue={username} />
                        <InfoInput
                            name="Bio"
                            multiline={true}
                            initValue={bio}
                        />
                        <InfoInput name="Website" initValue={website} />
                        <InfoInput name="Location" initValue={location} />
                        <InfoInput name="Pronouns" initValue="Add pronouns" />
                    </View>
                </View>
            </ScrollView>
        </MainContainer>
    );
}