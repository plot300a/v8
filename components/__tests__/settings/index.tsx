import * as React from 'react';
import {render, screen,} from '@testing-library/react-native';
import {InfoInput, ProfileInfo, TabHead} from '../../settings/index'



jest.mock('@expo/vector-icons', () => {
  const { View } = require('react-native');
  return {
    Ionicons: View,
  };
});


describe('Test TabHead component', () => {
    const title = 'Posts'
  it('Renders correctly', () => {
    render(<TabHead title={title} />)
    expect(screen.toJSON()).toMatchSnapshot();
  });
  it('Tab heading or title is displayed', () => {
    render(<TabHead title={title} />)
    expect(screen.getByTestId('tab-title')).toBeVisible();
  });
  it('Screen has the right heading or title', () => {
    render(<TabHead title={title} />)
    expect(screen.getByTestId('tab-title')).toHaveTextContent(title);
  });
});

describe('Test InfoInput component', () => {
    const name = 'Bio'
    const value = 'Hello, this is something small about me'
  it('Renders correctly', () => {
    render(<InfoInput name={name} initValue={value} />);
    expect(screen.getByTestId(`input-field-${name}`)).toBeVisible();
    expect(screen.getByTestId(`input-field-${name}`)).toHaveProp('accessibilityLabel',`Edit ${name}`)
    expect(screen.toJSON()).toMatchSnapshot();
  });
});

describe('Test ProfileInfo component', () => {
   // Some needed data
   const imageUrl = "",
   devImage = undefined,
   username="Alexander Moore",
   bio="Something little about user",
   website="www.versess.com",
   followers=534,
   folllowing=54,
   location="Accra, Ghana";
  it('Renders correctly', () => {
    render(
        <ProfileInfo  
            username={username}
            bio={bio}
            imageUrl={imageUrl}
            website={website}
            followers={followers}
            folllowing={folllowing}
            location={location}
            devImage={devImage}
        />
    );
    
    expect(screen.getByLabelText(`${username}'s profile image`)).toBeVisible()
    expect(screen.getByText(username)).toBeVisible()
    expect(screen.getByText(website)).toBeVisible()
    expect(screen.getByText(location)).toBeVisible()
  });
});
