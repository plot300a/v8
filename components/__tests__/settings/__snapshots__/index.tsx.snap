// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test InfoInput component Renders correctly 1`] = `
<View
  accessibilityLabel="Edit Bio"
  accessible={true}
  style={
    {
      "marginBottom": 25,
    }
  }
  testID="input-field-Bio"
>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
        },
        {
          "color": "#707070",
          "fontWeight": "bold",
          "marginBottom": 5,
        },
      ]
    }
  >
    Bio
  </Text>
  <TextInput
    style={
      [
        {
          "backgroundColor": "rgba(0,0,0,0)",
          "fontSize": 14,
          "marginBottom": 3,
          "marginTop": 10,
          "paddingBottom": 5,
          "width": "100%",
        },
        {
          "color": "#000000",
        },
      ]
    }
    value="Hello, this is something small about me"
  />
  <View
    style={
      {
        "backgroundColor": "#E3E3E3",
        "height": 1,
        "width": "100%",
      }
    }
  />
</View>
`;

exports[`Test TabHead component Renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "marginLeft": -10,
      "marginRight": 15,
      "opacity": 1,
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#000000",
          "fontSize": 14,
        },
        {
          "color": "#707070",
          "fontSize": 15,
          "fontWeight": "bold",
          "paddingBottom": 5,
        },
      ]
    }
    testID="tab-title"
  >
    Posts
  </Text>
</View>
`;
